crates/bdk/src/wallet/mod.rs

```diff
@@ -472,66 +460,91 @@ impl<D> Wallet<D> {
+            ConfirmationTime::Confirmed { height, time } => {
+                // anchor tx to checkpoint with lowest height that is >= position's height
+                let anchor = self
+                    .chain
+                    .blocks()
+                    .range(height..)
+                    .next()
+                    .ok_or(InsertTxError::ConfirmationHeightCannotBeGreaterThanTip {
+                        tip_height: self.chain.tip().map(|b| b.height),
+                        tx_height: height,
+                    })
+                    .map(|(&anchor_height, &anchor_hash)| ConfirmationTimeAnchor {
+                        anchor_block: BlockId {
+                            height: anchor_height,
+                            hash: anchor_hash,
+                        },
+                        confirmation_height: height,
+                        confirmation_time: time,
+                    })?;
+
+                (Some(anchor), None)
+            }
+            ConfirmationTime::Unconfirmed { last_seen } => (None, Some(last_seen)),
+        };
+
+        let changeset: ChangeSet = self.indexed_graph.insert_tx(&tx, anchor, last_seen).into();
+        let changed = !changeset.is_empty();
         self.persist.stage(changeset);
         Ok(changed)
     }
 
-    #[deprecated(note = "use Wallet::transactions instead")]
-    /// Deprecated. use `Wallet::transactions` instead.
-    pub fn list_transactions(&self, include_raw: bool) -> Vec<TransactionDetails> {
-        self.keychain_tracker
-            .chain()
-            .txids()
-            .map(|&(_, txid)| self.get_tx(txid, include_raw).expect("must exist"))
-            .collect()
-    }
-
-    /// Iterate over the transactions in the wallet in order of ascending confirmation time with
-    /// unconfirmed transactions last.
+    /// Iterate over the transactions in the wallet.
     pub fn transactions(
         &self,
-    ) -> impl DoubleEndedIterator<Item = (ConfirmationTime, &Transaction)> + '_ {
-        self.keychain_tracker
-            .chain_graph()
-            .transactions_in_chain()
-            .map(|(pos, tx)| (*pos, tx))
+    ) -> impl Iterator<Item = CanonicalTx<'_, Transaction, ConfirmationTimeAnchor>> + '_ {
+        self.indexed_graph
+            .graph()
+            .list_chain_txs(&self.chain, self.chain.tip().unwrap_or_default())
     }
 
     /// Return the balance, separated into available, trusted-pending, untrusted-pending and immature
     /// values.
     pub fn get_balance(&self) -> Balance {
-        self.keychain_tracker.balance(|keychain| match keychain {
-            KeychainKind::External => false,
-            KeychainKind::Internal => true,
-        })
+        self.indexed_graph.graph().balance(
+            &self.chain,
+            self.chain.tip().unwrap_or_default(),
+            self.indexed_graph.index.outpoints().iter().cloned(),
+            |&(k, _), _| k == KeychainKind::Internal,
```

Comment:
This reminds me that I had a nice chat w/ some friends regarding the meaning of "trusted" and "untrusted" keychain and how it doesn't really make sense as a distinction. I will post a message on discord explaining (eventually 😅 ) or directly update the balance issue we already have 

But the TL;DW is: addresses are public anyways, everyone can send money to our change addresses, so you should distinguish between trusted and untrusted based on whether you did a certain tx or not, not based on the keychain used